core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala

```diff
@@ -0,0 +1,347 @@
+      )
+    }
+
+    result
+  }
+
+  private def convertToUnsigned(targetColumnInfo: TiColumnInfo, value: AnyRef): Object = {
+    val lowerBound = 0L
+    val upperBound = integerUnsignedUpperBound(targetColumnInfo.getType)
+
+    val result: java.lang.Long = value match {
+      case v: java.lang.Boolean => if (v) 1L else 0L
+      case v: java.lang.Byte    => v.longValue()
+      case v: java.lang.Short   => v.longValue()
+      case v: java.lang.Integer => v.longValue()
+      case v: java.lang.Long    => v.longValue()
+      case v: java.lang.Float   => floatToLong(v)
+      case v: java.lang.Double  => doubleToLong(v)
+      case v: String            => stringToLong(v)
+      // TODO: case v: java.math.BigDecimal => v.longValue()
+      // TODO: support following types
+      //case v: java.sql.Date              =>
+      //case v: java.sql.Timestamp         =>
+      //case v: Array[String]              =>
+      //case v: scala.collection.Seq[_]    =>
+      //case v: scala.collection.Map[_, _] =>
+      //case v: org.apache.spark.sql.Row   =>
+      case _ =>
+        throw new TiBatchWriteException(
+          s"do not support converting from ${value.getClass} to column type: ${targetColumnInfo.getType}"
+        )
+    }
+
+    if (result < lowerBound) {
+      throw new TiBatchWriteException(
+        s"data $value < lowerBound $lowerBound"
+      )
+    }
+
+    if (java.lang.Long.compareUnsigned(result, upperBound) > 0) {
+      throw new TiBatchWriteException(
+        s"data $value > upperBound $upperBound"
+      )
+    }
+
+    result
+  }
+
+  private def convertToReal(targetColumnInfo: TiColumnInfo, value: AnyRef): Object = {
+    val result: java.lang.Double = value match {
+      case v: java.lang.Boolean => if (v) 1d else 0d
+      case v: java.lang.Byte    => v.doubleValue()
+      case v: java.lang.Short   => v.doubleValue()
+      case v: java.lang.Integer => v.doubleValue()
+      case v: java.lang.Long    => v.doubleValue()
+      case v: java.lang.Float   => v.doubleValue()
+      case v: java.lang.Double  => v
+      case v: String            => stringToDouble(v)
+      // TODO: case v: java.math.BigDecimal => v.doubleValue()
+      // TODO: support following types
+      //case v: java.sql.Date              =>
+      //case v: java.sql.Timestamp         =>
+      //case v: Array[String]              =>
+      //case v: scala.collection.Seq[_]    =>
+      //case v: scala.collection.Map[_, _] =>
+      //case v: org.apache.spark.sql.Row   =>
+      case _ =>
+        throw new TiBatchWriteException(
+          s"do not support converting from ${value.getClass} to column type: ${targetColumnInfo.getType}"
+        )
+    }
+
+    targetColumnInfo.getType.getType match {
+      case MySQLType.TypeFloat =>
+        val lowerBound = -java.lang.Float.MAX_VALUE
+        val upperBound = java.lang.Float.MAX_VALUE
+
+        if (java.lang.Double.compare(result, upperBound) > 0) {
```

Comment:
I doubt whether we should use Double.compare to compare double values. It should be better to use EPSILON in order to reduce the impact of precision loss.

Reply:
fix